packages/firestore/src/remote/bloom_filter.ts

```diff
@@ -0,0 +1,84 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Md5, Integer } from '@firebase/webchannel-wrapper';
+
+import { debugAssert } from '../util/assert';
+
+export class BloomFilter {
+  private readonly bitSize: number;
+
+  constructor(
+    private readonly bitmap: Uint8Array,
+    padding: number,
+    private readonly hashCount: number
+  ) {
+    this.bitSize = this.bitmap.length * 8 - padding;
+    if (this.bitSize === 0) {
+      debugAssert(
+        this.hashCount === 0 && padding === 0,
+        'A valid empty bloom filter should have all 3 fields empty.'
+      );
+    } else {
+      debugAssert(padding >= 0, 'Padding is negative.');
+      debugAssert(this.bitSize > 0, 'Bitmap size is negative.');
+      debugAssert(this.hashCount > 0, 'Hash count is 0 or negative');
+    }
+  }
+
+  getBitSize(): number {
+    return this.bitSize;
+  }
+
+  mightContain(document: string): boolean {
+    // Empty bitmap should always return false on membership check
+    if (this.bitSize === 0) {
+      return false;
+    }
+
+    // Hash the string using md5
+    const md5 = new Md5();
+    md5.update(document);
```

Comment:
I believe the string specified to MD5's `update()` method needs to be encoded into UTF-8 bytes. This won't probably make a difference when using only standard ASCII characters; however, it will make a huge difference when the string has non-standard characters, like é and 😂. Please work with Xijiao to add a golden test for this case (if one doesn't exist) because we absolutely must be certain that extended characters work correctly.

You can use the `TextEncoder` class to encode a string into UTF-8. Note that we have a "helper" method to get instances of TextEncoder: https://github.com/firebase/firebase-js-sdk/blob/fde5adf638dae4714ff7f25c75e475344907e05e/packages/firestore/src/core/firestore_client.ts#L44

Reply:
Excellent point! Just tested, with and with out UTF-8 encoding, non standard characters will return different hash value. Will confirm with Xijiao on Monday.